content/algorithms/planarity/planarity.md

```diff
@@ -0,0 +1,939 @@
+3. *Conflict Detection:*  
+The algorithm tracks embedding constraints between edges. If it detects a conflict (e.g., an edge that can't be placed without violating planarity), it concludes the graph is non-planar.
+4. *Constructs a Planar Embedding or Finds Obstruction:*  
+If the graph is planar, it returns a combinatorial embedding (rotation system). If not, it can identify a Kuratowski subgraph (e.g., K₅ or K₃,₃).
+
+
+
+
+
+
+```python
+from functools import singledispatch
+```
+
+
+```python
+from collections import defaultdict
+import networkx as nx
+import numpy as np
+import matplotlib.pyplot as plt
+```
+
+
+```python
+__all__ = ["check_planarity", "is_planar", "PlanarEmbedding"]
+```
+
+## Parameters
+
+G: NetworkX graph
+counterexample : bool
+A Kuratowski subgraph (to prove non planarity) is only returned if set to true.
+
+
+What is ***Kuratowski subgraph*** ?
+
+A Kuratowski subgraph is a specific subgraph that serves as evidence that a graph is not planar. According to Kuratowski’s Theorem, a graph is non-planar if and only if it contains a subgraph that is a subdivision (i.e., homeomorphic copy) of either **K₅ (the complete graph on 5 vertices)** or **K₃,₃ (the complete bipartite graph with partitions of size 3)**. These subgraphs are called Kuratowski subgraphs.
+
+They may not appear in their exact form but can be present in a stretched-out version where edges are replaced by paths. NetworkX uses this principle internally: *if check_planarity() returns False, it indicates the existence of a Kuratowski subgraph obstructing planar embedding.*
+
+---
+
+
+
+## Returns
+
+The function returns a tuple of the form:
+
+
+*   is_planar (bool):
+  *   True if the graph is planar.
+  *   False if the graph is non-planar.
+
+*   certificate (NetworkX graph):
+  *   If is_planar is True, certificate will be a PlanarEmbedding object representing a valid planar embedding of the graph.
+  *   If is_planar is False, certificate will be a Kuratowski subgraph, which is a subgraph of the original graph that proves the graph is non-planar by being a homeomorphic copy of K₅ or K₃,₃.
+
+---
+
+
+
+
+```python
+from networkx.algorithms.planarity import LRPlanarity
+```
+
+
+```python
+def is_planar(G):
+    return check_planarity(G, counterexample=False)[0]
+```
+
+
+```python
+def check_planarity(G, counterexample=False):
+    planarity_state = LRPlanarity(G)
+    embedding = planarity_state.lr_planarity()
+    if embedding is None:
+        # graph is not planar
+        if counterexample:
+            return False, get_counterexample(G)
+        else:
+            return False, None
+    else:
+        # graph is planar
+        return True, embedding
+```
+
+
+```python
+#Example 1
+# Load GraphML file
+G = nx.read_graphml("./data/planar_graph.graphml")
```

Comment:
You can create the graph in code for the small graphs in this notebook.  That is more readable because they can see the nodes and edges without all the code and distraction of loading the file.
```python
G = nx.Graph([(0, 1), (1, 2), (2, 3), (0, 2)])
```